kratos/containers/master_slave_relation.h

```diff
@@ -0,0 +1,201 @@
+ *
+ *
+ *  IMPORTANT : This is not seen by the user. This is a helper data structure which is created in the builder and solver and destroyed there.
+ *
+ * @author Aditya Ghantasala
+ */
+class MasterSlaveRelation : public IndexedObject
+{
+  public:
+    typedef std::size_t IndexType;
+    typedef Matrix MatrixType;
+    typedef Vector VectorType;
+    typedef std::vector<IndexType> EquationIdVectorType;
+    KRATOS_CLASS_POINTER_DEFINITION(MasterSlaveRelation);
+
+    // empty constructor and methods to add master and slave independently.
+    MasterSlaveRelation() : IndexedObject(0)
+    {
+        SetConstant(0.0);
+        //SetConstantUpdate(0.0);
+    }
+
+
+    MasterSlaveRelation(IndexType const &rSlaveEquationId) : IndexedObject(rSlaveEquationId)
+    {
+        SetConstant(0.0);
+        //SetConstantUpdate(0.0);
+    }
+
+    void SetConstant(double Constant) { mConstant = Constant; }
+    //void SetConstantUpdate(double ConstantUpdate) { mConstantUpdate = ConstantUpdate; }
+    double Constant() const { return mConstant; }
+    //double ConstantUpdate() const { return mConstantUpdate; }
+    IndexType SlaveEquationId() const { return this->Id(); }
+
+
+
+    // Get number of masters for this slave
+    std::size_t GetNumberOfMasters() const
+    {
+        return mMasterDataSet.size();
+    }
+
+
+    /**
+     * this determines the master equation IDs connected to this constraint
+     * @param rResult the elemental equation ID vector
+     * @param rCurrentProcessInfo the current process info instance
+     */
+    virtual void EquationIdVector(EquationIdVectorType& rSlaveEquationIds,
+                                  EquationIdVectorType& rMasterEquationIds,
+                                  ProcessInfo& rCurrentProcessInfo)
+    {
+        if (rSlaveEquationIds.size() != 0)
+            rSlaveEquationIds.resize(1);
+
+        if (rMasterEquationIds.size() != 0)
+            rMasterEquationIds.resize(this->GetNumberOfMasters());
+
+        rSlaveEquationIds[0] = this->SlaveEquationId();
+
+        auto master_it = mMasterDataSet.begin();
+        for (IndexType i=0; i<this->GetNumberOfMasters(); i++)
+            rMasterEquationIds[i] = (*(++master_it)).first;
+    }
+
+    /**
+     * this is called during the assembling process in order
+     * to calculate all elemental contributions to the global system
+     * matrix and the right hand side
+     * @param rTransformationMatrix the elemental left hand side matrix
+     * @param rConstant the elemental right hand side
+     * @param rCurrentProcessInfo the current process info instance
+     */
+    virtual void CalculateLocalSystem(MatrixType& rTransformationMatrix,
+                                      VectorType& rConstantVector,
+                                      ProcessInfo& rCurrentProcessInfo)
+    {
+      if (rTransformationMatrix.size1() != 0)
```

Comment:
! this should include the case of multiple master dofs...

Reply:
i mean multiple slaves and multiple masters

Reply:
Yes ... I will add this constructor also.